lib/Sema/DerivedConformanceEquatableHashable.cpp

```diff
@@ -32,26 +32,181 @@
 using namespace swift;
 using namespace DerivedConformance;
 
+/// Returns true if the given type conforms to the protocol.
+/// \p type The type whose conformance should be checked.
+/// \p protocol The protocol being requested.
+/// \return True if the type conforms to the protocol.
+static bool typeConformsToProtocol(TypeChecker &tc, DeclContext *context,
+                                   Type type, ProtocolDecl *protocol) {
+  return tc.conformsToProtocol(type, protocol, context,
+                               ConformanceCheckFlags::Used).hasValue();
+}
+
+/// Returns true if, for every element of the given enum, it either has no
+/// associated values or all of them conform to a protocol.
+/// \p theEnum The enum whose elements and associated values should be checked.
+/// \p protocol The protocol being requested.
+/// \return True if all associated values of all elements of the enum conform.
+bool allAssociatedValuesConformToProtocol(TypeChecker &tc, EnumDecl *theEnum,
+                                          ProtocolDecl *protocol) {
+  auto declContext = theEnum->getDeclContext();
+
+  for (auto elt : theEnum->getAllElements()) {
+    auto argumentTypeLoc = elt->getArgumentTypeLoc();
+    if (argumentTypeLoc.isNull())
+      continue;
+    auto argumentType = argumentTypeLoc.getType();
+    if (auto tupleType = argumentType->getAs<TupleType>()) {
+      // One associated value with a label or multiple associated values
+      // (labeled or unlabeled) are tuple types.
+      for (auto tupleElementType : tupleType->getElementTypes()) {
+        if (!typeConformsToProtocol(tc, declContext, tupleElementType,
+                                    protocol)) {
+          return false;
+        }
+      }
+    } else {
+      // One associated value with no label is represented as a paren type.
+      auto actualType = argumentType->getWithoutParens();
+      if (!typeConformsToProtocol(tc, declContext, actualType, protocol)) {
```

Comment:
Nit: This, and the call above it, can be used as the return value of this branch.